pkg/translator/prometheus/translator.go

```diff
@@ -0,0 +1,104 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package prometheus // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/prometheus"
+
+import (
+	"strings"
+	"unicode"
+
+	"go.opentelemetry.io/collector/pdata/pmetric"
+)
+
+type Translator interface {
+	TranslateAttribute(string) string
+	TranslateMetric(metric pmetric.Metric, namespace string, addMetricSuffixes bool) string
+}
+
+type UTF8AllowedTranslator struct{}
+
+func (UTF8AllowedTranslator) TranslateAttribute(attribute string) string {
+	return addKeyPrefixIfNeeded(attribute)
+}
+
+func (UTF8AllowedTranslator) TranslateMetric(metric pmetric.Metric, namespace string, addMetricSuffixes bool) string {
+	tokens := make([]string, 0)
+	if namespace != "" {
+		tokens = append(tokens, namespace)
+	}
+	tokens = append(tokens, metric.Name())
+
+	if addMetricSuffixes && normalizeNameGate.IsEnabled() {
```

Comment:
This part here feels awkward. Two feature flags that fundamentally goes against each other.

NormalizeNameGate is used tell if the user wants Prometheus Normalization or not, which means utf8 turns into underscore, units become part of the metric name and type suffixes are also added.

If NormalizeName and AllowUTF8 are both enabled at the same time, I guess we implement partial normalization (allow utf8 but still add suffixes)? When adding the suffixes, should we use dots or underscores to separate from the metric name?

I'm inclining to not have yet another feature gate and just stop removing forbidden runes when `normalizeNameGate`  is disabled 🤔

Reply:
The normalization of Otel metrics when exporting to Prometheus performs 3 operations:

1. Remove unsupported runes in Prometheus
2. Append the unit as a suffix
3. Follow some additional Prometheus conventions (`_total` suffix for counters, etc.)

Allowing UTF8 means we can skip step 1 in the above list, and only this one.

Reply:
That means that we'll have different behaviors for `NormalizeNameGate`, right? Feels weird to me, I'd prefer that they don't interact with each other 😬 

No strong opinions though, since NormalizeNameGate is on by default(beta) and the new gate will be off by default (alpha) we won't be introducing breaking changes immediately

Reply:
To me we should have:

| normalizeNameGate | allowUTF8 | Operations |
|---|---|---|
| `false` | `false` | Remove unsupported runes, don't follow Prometheus conventions |
| `false` | `true` | Do nothing, leave metric names as is |
| `true` | `false` | Current default behavior: remove unsupported runes in Prometheus 2.x, and follow Prometheus conventions, like unit suffixes, `_total` suffix, `_ratio`, etc.)
| `true` | `true` | Let all runes in place **AND** follow Prometheus conventions: unit suffixes, `_total` , `_ratio`, etc. |

Make sure the code behaves as in the above table and we should be fine! 😊

Reply:
I've just added a test that verifies their interaction :)